### ISSUE TITLE
linter error

### ISSUE THREAD

**Issue by jakoch** [rhysd/actionlint#519]
## Input

Given this workflow.yml
```
name: Lint

on: [push, pull_request]

jobs:
  lint:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Advanded Linting Devices Inc.
        run: |
          sudo apt-get update
          sudo apt-get install -y advanced-linting-device

      - name: Set locale
        run: |
          export LC_ALL=C.UTF-8

        - name: COMMON → Lint line endings (show files with DOS LE and EOL errors)
          run: |
            find -name '*.*' -print0 | xargs -0 dos2unix -ice
```

## Expected Output
Yay! No error was detected.

## Actual Output

- `could not parse as YAML: yaml: line 16: did not find expected key [syntax-check]`

## Analysis

The line `find -name '*.*' -print0 | xargs -0 dos2unix -ice` causes the linter to assume an invalid `could not parse as YAML: yaml: line 16: did not find expected key [syntax-check]` error.

Just for info where Line 16 is:

```
      - name: Advanded Linting Devices Inc.
        run: |
          sudo apt-get update
          sudo apt-get install -y advanced-linting-device
This is L16
      - name: Set locale
        run: |
          export LC_ALL=C.UTF-8
```

Which means the linter is fully derailed.

## References

Error Example and Error Exclusion for ref.

- [Playground Example no error](https://rhysd.github.io/actionlint/#eJx1jkFLAzEQhe/5Fe8HmHrxIAHBsiIIe7OeREqaDO1qTNKdmUXBH2+2LT0Uehoe7715nynZ4b0q725QNaX1SHsllg9jPsuGnQHSkGW+wKiZ7VzQjWZRm7y06MFiocrHFGCR/Tc5dDsKX0UFoUQ6eYAysYMPMpTMt+GUeZzuzEV9GSefI0X0jWDIWzzRNARivOSwOL9rUA5/Z9lQNBb4KnZLAq2xQV5zh8ziU4L9hZ/HAkWbjmM2HsYumV5bK5XgE10HoJ9aRkHfrZd9/9At3lbP9v4f6UNpvA==)
- [Playground Error](https://rhysd.github.io/actionlint/#eJx1kMFKAzEURff5irurFlOLuJABwdJWEKZ2UV2JlDR5dqIxGSdJW6FrP8BP9EvMTMsgla6G8O6dc95jVrxThlzbwJizGZ7K6IszlNGYeUUfkXx4ZuzVLXzGAJNy9ReoovW8LsRFtCFyI0KKNiMfqPS7FMCxIwwLkm8uBkinaD8DoiefQcignfXncp+5WV2yg/pArYRVpBpTbZcY0UpL8rizstf+Lkll2LbPpBKVgygDX1JALFWSPDbV1gdhDPgnRA2TpLjZwbhqYIdOs9QyTgpDxwVoU7oqIB/OB3l+Pew9PtzyK9YG2utMJ5PpPX6+vpsF6zsTyKoE9zjxhVvjRZu07lqHAqPpDPkY6SAYT3NQVbnKn/6h/tNAqqc0r2nodHvdDnhZJVIfW2xElSi8D+X8RbR6A562/QXlAJyQ)

**Comment by ghost** [comment]
For fellow Googlers:

THis cryptic error also happens when using brackets: `[foo]` which actionlint doesn't like at all (even though GA works)

**Comment by muzimuzhi** [comment]
@jakoch As a comparison, the pyyaml-based YAML parser throws two errors on your YAML example ([link](https://yaml-online-parser.appspot.com/?yaml=name%3A+Lint%0A%0Aon%3A+%5Bpush%2C+pull_request%5D%0A%0Ajobs%3A%0A++lint%3A%0A++++runs-on%3A+ubuntu-latest%0A++++steps%3A%0A++++++-+name%3A+Checkout+code%0A++++++++uses%3A+actions%2Fcheckout%40v4%0A%0A++++++-+name%3A+Advanded+Linting+Devices+Inc.%0A++++++++run%3A+%7C%0A++++++++++sudo+apt-get+update%0A++++++++++sudo+apt-get+install+-y+advanced-linting-device%0A%0A++++++-+name%3A+Set+locale%0A++++++++run%3A+%7C%0A++++++++++export+LC_ALL%3DC.UTF-8%0A%0A++++++++-+name%3A+COMMON+%E2%86%92+Lint+line+endings+(show+files+with+DOS+LE+and+EOL+errors)%0A++++++++++run%3A+%7C%0A++++++++++++find+-name+%27*.*%27+-print0+%7C+xargs+-0+dos2unix+-ice&type=json)):

```
ERROR:

while parsing a block mapping
  in "<unicode string>", line 17, column 9:
          - name: Set locale
            ^
expected <block end>, but found '-'
  in "<unicode string>", line 21, column 9:
            - name: COMMON → Lint line endin ... 
            ^
```

@riva-infinex Could you provide a concrete example?

**Comment by ghost** [comment]
@muzimuzhi cryptic error https://rhysd.github.io/actionlint/#eJxTVshIzcnJV+TiykvMTbVSiHbLz49VSEos4gIAZuYHww==

**Comment by muzimuzhi** [comment]
@riva-infinex `name: [Foo] bar` is an invalid YAML. https://yaml-online-parser.appspot.com/?yaml=name%3A+%5BFoo%5D+bar&type=json

`[...]` represents an inline sequence (technically called "[flow sequence](https://yaml.org/spec/1.2.2/#flow-sequences)"), which cannot be followed by a string on the same line. To specify a string which starts with `[`, quoting the whole string is a common workaround.

```yaml
# single-quoted
name: '[Foo] bar'
# double-quoted
name2: "[Foo] bar"
```

https://yaml-online-parser.appspot.com/?yaml=name%3A+%27%5BFoo%5D+bar%27%0Aname2%3A+%22%5BFoo%5D+bar%22&type=json

_Update:_ If the GitHub Actions accepts such YAML snippet, then it's YAML parser might not be strict enough.

**Comment by ghost** [comment]
@muzimuzhi I completely agree! My comment is about the cryptic error message:

1. (playground above) it complains about line 2 when the error is on line 3
2. the message "did no find expected key" doesn't mention anything about brackets or flow sequences

**Comment by muzimuzhi** [comment]
@riva-infinex The cryptic error message "did not find expected key" was provided by the YAML parser `gopkg.in/yaml.v3 v3.0.1` used by `actionlint`, so it's an upstream issue. See reproduction steps at the end.

https://github.com/rhysd/actionlint/blob/2ab3a12c7848f6c15faca9a92612ef4261d0e370/go.mod#L16

The shifted line number seems to be caused by `actionlint`.

Reproduction of "did not find expected key" error message:

- open https://pkg.go.dev/gopkg.in/yaml.v3@v3.0.1#example-Unmarshal-Embedded
- Edit the embedded example
  ```diff
   var data = `
  -a: a string x from struct A
  +a: [Foo] bar
   b: a string x from struct B
   `
  ```
- Click the lower-right "Run" button, wait, and get the error message
  ```
  2009/11/10 23:00:00 cannot unmarshal data: yaml: line 1: did not find expected key
  ```

Unfortunately, the source repo of the Go package `gopkg.in/yaml` has been archived on Apr 2, 2025. See https://github.com/go-yaml/yaml.

**Comment by jakoch** [comment]
Thank you for taking the time to analyze this.
I just checked [the dependencies of yq](https://github.com/mikefarah/yq/blob/master/go.mod?utm_source=chatgpt.com) and noticed it includes [goccy/go-yaml](https://github.com/goccy/go-yaml?utm_source=chatgpt.com). This might serve as a potential replacement package.
